File: lang/java/src/fun/vac/algorithms/sorting/QuickSort.java
package fun.vac.algorithms.sorting;

public class QuickSort {

    private QuickSort() {}

    public static <T extends Comparable<T>> void partition(T[] arr, int lo, int hi) {
        if (lo == hi) {
            return;
        }

        T pivot = arr[lo];

        int left = lo, right = hi - 1;

        while (left != right) {
            for (int i = right; i > left; i--) {
                if (arr[i].compareTo(pivot) < 0) {
                    arr[left] = arr[i];
                    arr[i] = pivot;
                    break;
                }
                right -= 1;
            }
            for (int i = left; i < right; i++) {
                if (arr[i].compareTo(pivot) > 0) {
                    arr[right] = arr[i];
                    arr[i] = pivot;
                    break;
                }
                left += 1;
            }
        }

        partition(arr, lo, left);
        partition(arr, left + 1, hi);
    }

    public static <T extends Comparable<T>> void sort(T[] arr) {
        partition(arr, 0, arr.length);
    }
}

File: lang/ruby/src/fun/vac/algorithms/sorting/QuickSort.rb
module QuickSort

    def self.partition(arr, lo, hi)
        if lo == hi
            return
        end

        pivot = arr[lo]

        left, right = lo, hi - 1

        until left == right
            i = right
            while i > left
                if arr[i] < pivot
                    arr[left] = arr[i]
                    arr[i] = pivot
                    break
                end
                right -= 1
                i -= 1
            end
            i = left
            while i < right
                if arr[i] > pivot
                    arr[right] = arr[i]
                    arr[i] = pivot
                    break
                end
                left += 1
                i += 1
            end
        end

        partition(arr, lo, left)
        partition(arr, left + 1, hi)
    end

    def self.sort(arr)
        partition(arr, 0, arr.length)
    end
end
